File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        // have some kind of container that keeps track of the letters in the current
        // substring
        // have a sliding window alg. that adds on whenever we have a character that is 
        // not a duplicate and whenever it is a duplicate, keep incrementing the start
        // counter until we get rid of the duplicate

        int start = 0, end = 0;
        Set<Character> container = new HashSet<>();
        int maxLength = 0;
        
        for(end = 0; end < s.length(); end++) {
            if(!container.contains(s.charAt(end))) {
                container.add(s.charAt(end));
            } else {
                while(s.charAt(start) != s.charAt(end)) {
                    container.remove(s.charAt(start));
                    start++;
                }

                start++;
            }

            maxLength = Math.max(maxLength, end - start + 1);
        }

        return maxLength;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left = 0
        m = set()
        ret = 0

        for right in range(len(s)):
            curr = s[right]

            while curr in m:
                m.remove(s[left])
                left+= 1

            m.add(s[right])
            ret = max(ret, right - left + 1)
        
        return ret

                